[SE-0225] Implementation of isMultiple for BinaryInteger.

stephentyrone opened the pull request:
Implementation tracking proposal for SE-0225.

Preliminary discussion here: https://forums.swift.org/t/even-and-odd-integers/11774/129
Proposal here: https://github.com/apple/swift-evolution/pull/891

Final decision: https://forums.swift.org/t/accepted-with-modifications-se-0225-adding-iseven-isodd-ismultiple-to-binaryinteger/15689

stephentyrone commented:
@swift-ci please test.

unnamedd commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3132,6 +3158,14 @@ extension UnsignedInteger where Self : FixedWidthInteger {
+  @inlinable
+  public func isMultiple(of other: Self) -> Bool {
+    // Nothing but zero is a multiple of zero.
+    if other == 0 { return self == 0 }
```
Why not use of `guard` for an early return?

stephentyrone replied:
In this case I think it reads a lot more clearly as an `if`, because it avoids the double-negative.

unnamedd replied:
Make sense! @stephentyrone thanks for your response 😉

unnamedd commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3241,6 +3275,16 @@ extension SignedInteger where Self : FixedWidthInteger {
+  @inlinable
+  public func isMultiple(of other: Self) -> Bool {
+    // Nothing but zero is a multiple of zero.
+    if other == 0 { return self == 0 }
```
Don't you consider to use `guard` for this early return?

stephentyrone commented:
@swift-ci Please smoke test

stephentyrone commented on test/Prototypes/DoubleWidth.swift.gyb:
```diff
@@ -464,6 +463,12 @@ extension DoubleWidth : FixedWidthInteger {
     return (quotient_, remainder_)
   }
 
+  public func isMultiple(of other: DoubleWidth) -> Bool {
```
We can't piggyback on the default implementation here because `DoubleWidth` doesn't unconditionally conform to `SignedInteger` or `UnsignedInteger` -- it's one or the other, but we don't know which.

stephentyrone commented:
@swift-ci Please smoke test.

stephentyrone committed:
Implement SE-0225 (BinaryInteger.isMultiple(of:))

A default implementation is provided for FixedWidthInteger. Very basic test coverage included as well.

stephentyrone commented:
@swift-ci please benchmark

xwu commented on stdlib/public/core/Integers.swift:
```diff
@@ -2755,6 +2770,16 @@ extension FixedWidthInteger {
+    // Nothing but zero is a multiple of zero.
+    if other == 0 { return self == 0 }
+    // Special case to avoid overflow on .min / -1 for signed types.
+    if Self.isSigned && self == .min && other == -1 { return true }
```
I'd be curious if performance changes if the two conditionals were replaced with a check of the magnitude, i.e. `return other.magnitude <= 1 || self % other == 0`...

stephentyrone replied:
For concrete types, there should basically be no difference in the face of optimization. We can actually re-arrange this slightly to have only a single test on the fast path, but the first thing is to get the API in.

stephentyrone committed:
Minor simplificiation.

stephentyrone commented:
@swift-ci please smoke test
